parity-clib/src/python.rs

```diff
@@ -0,0 +1,219 @@
+	fn drop(&mut self) {
+		self.unsubscribe()
+	}
+}
+
+/// A Python callable which accepts a String value representing a RPC result or subscription event
+struct PythonCallback(PyObject);
+
+impl PythonCallback {
+	fn new(callback: PyObject) -> Self {
+		Self(callback)
+	}
+}
+
+impl Callback for PythonCallback {
+	fn call(&self, msg: &str) {
+		// Assume this is called from Rust, acquire GIL
+		let gil = Python::acquire_gil();
+		let py = gil.python();
+
+		// Run our callback, and re-raise any exceptions that occur
+		if let Err(e) = self.0.call1(py, msg) {
+			e.restore(py);
+		}
+	}
+}
+
+/// Generate a Parity config given a list of CLI options
+#[pyfunction]
+fn config_from_cli(_py: Python<'_>, cli: &PyList) -> PyResult<ConfigHandle> {
+	let cli_len = cli.len();
+
+	let opts_strings = cli
+		.into_iter()
+		.map(|o| o.extract())
+		.collect::<PyResult<Vec<String>>>()?;
+
+	let opts = opts_strings
+		.iter()
+		.map(|s: &String| s.as_ptr() as *const i8)
+		.collect::<Vec<_>>();
+	let opts_lens = opts_strings
+		.iter()
+		.map(|s| s.as_bytes().len())
+		.collect::<Vec<_>>();
+
+	let mut out = ptr::null_mut();
+	unsafe {
+		match parity_config_from_cli(opts.as_ptr(), opts_lens.as_ptr(), cli_len, &mut out) {
+			0 => Ok(ConfigHandle { inner: out }),
+			_ => Err(ValueError::py_err("failed to create config object")),
+		}
+	}
+}
+
+/// Create a running Parity instance given a config
+#[pyfunction]
+unsafe fn build(
+	_py: Python<'_>,
+	config: &mut ConfigHandle,
+	logger_mode: &str,
+	logger_file: &str,
+) -> PyResult<ParityHandle> {
+	let mut params = ParityParams {
+		configuration: config.inner,
+		..mem::zeroed()
+	};
+
+	parity_set_logger(
+		logger_mode.as_ptr(),
+		logger_mode.as_bytes().len(),
+		logger_file.as_ptr(),
+		logger_file.as_bytes().len(),
+		&mut params.logger,
+	);
+
+	let mut out = ptr::null_mut();
+	let ret = match parity_start(&params, &mut out) {
+		0 => Ok(ParityHandle { inner: out }),
+		_ => Err(RuntimeError::py_err("failed to start Parity")),
+	};
+
+	// Ensure we don't double-free config
+	config.inner = ptr::null_mut();
+	ret
+}
+
+/// Perform a RPC query against a running Parity instance, invoking callback on completion
+#[pyfunction]
+fn rpc_query(
+	_py: Python<'_>,
+	parity: &ParityHandle,
+	rpc: &str,
+	timeout_ms: u64,
+	callback: PyObject,
+) -> PyResult<()> {
+	if parity.inner == ptr::null_mut() {
+		return Err(ValueError::py_err("Parity execution completed"));
+	}
+
+	let client = unsafe { &*(parity.inner as *const RunningClient) };
+	let callback = Arc::new(PythonCallback::new(callback));
+	parity_rpc_worker(client, rpc, callback, timeout_ms);
+	Ok(())
+}
+
+/// Subscribe to a websocket event, invoking callback when event received
+#[pyfunction]
+fn subscribe_ws(
+	_py: Python<'_>,
+	parity: &ParityHandle,
+	rpc: &str,
+	callback: PyObject,
+) -> PyResult<SubscriptionHandle> {
+	if parity.inner == ptr::null_mut() {
+		return Err(ValueError::py_err("Parity execution completed"));
```

Comment:
hrm, this means that `ParityClient` has been shutdown or it failed during init right?

Can you change the error message something like: "Attempt to subscribe to WebSocket when ParityClient is not running"?